### ISSUE TITLE
GraphQL Lists deeper down

### ISSUE THREAD

**Issue by fragglebob** [DataDog/dd-trace-js#169]
How deep down is GraphQL supposed to instrument?

With a query like the following only these fields `friends`,  `friends.0.name`,  `friends.0.pets`,  `friends.1.name` and `friends.1.pets` get instrumented. Should all the fields on the types pet and color, get instrumented too? 

```graphql
{
  friends {
    name
    pets {
      name
      colors {
        code
      }
    }
  }
}
```

**Comment by rochdev** [comment]
Lists have been problematic and will be fixed in the next release.

The new behavior for the integration in general will be:

• every operation, either query or mutation, gets a span
• every field gets a span that includes the duration of its resolver and all child fields, but only if is has an associated resolver
• every resolver gets a span that includes only its own duration
• any field that doesn't have an explicit resolver, either on a programmatic schema or on the rootValue, will not be instrumented

So basically, any field that matches those criteria would get instrumented. For example, if you have a resolver on `friends` and `code`, the fields `friends` and `friends.0.pets.0.colors.0.code` would be instrumented.

The reasoning behind this behavior is that for example, if there is a list of 1000 friends, each of which has 3 pets with 3 colors each, you would get 9000 spans for a single operation. This would both incur a very high overhead and be too verbose in the UI.

Will this work for your use case?

**Comment by rochdev** [comment]
I'll take a look at the recursion error.

Otherwise, does the behavior mentioned above match what you are seeing? And does it make sense for your use case?

**Comment by fragglebob** [comment]
I've made a fork and I've got these things sorted in it, I didn't want to make a PR though until I knew you want to instrument that deep.

**Comment by rochdev** [comment]
PR for the recursion problem: https://github.com/DataDog/dd-trace-js/pull/170

**Comment by rochdev** [comment]
Just saw your branch, I guess you beat me to it :)

**Comment by rochdev** [comment]
Feel free to open the PR, we do want to instrument as deep as possible whenever there are custom resolvers. If this causes a cardinality issue we will simply add a configuration option.

**Comment by fragglebob** [comment]
Ok will do. 😉

**Comment by fragglebob** [comment]
Okay my PR for both issues is here: #171

**Comment by rochdev** [comment]
Fixed by #171
